### PR TITLE
[Boost][JP-Ellis&Artem] Update Boost version to 1.79

### LINKED ISSUE #1489: Boost dependency can be (mostly) upgraded
Previously, there was an incompatibility between Boost's `noinline` and what was provided within the HIP compiler which was initially reported in #658.  Since then, the Boost library has been patched through boostorg/config#393 which was included in Boost version 1.78.  As a result, I have tried again to see if I can avoid having to recompile Boost and use the system version and it mostly works.

In particular, the compilation works fine for all files except for the following 9 files:
- `src/find_db.cpp`
- `src/binary_cache.cpp`
- `src/sqlite_db.cpp`
- `src/solver/conv_asm_3x3u.cpp`
- `src/ocl/gcn_asm_utils.cpp`
- `src/ocl/mloNeuron.cpp`
- `src/ocl/mloPooling.cpp`
- `src/ocl/mloNorm.cpp`
- `src/hip/hip_build_utils.cpp`

The error message is similar to the original issue:

```text
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/find_db.cpp:29:
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/include/miopen/handle.hpp:30:
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/include/miopen/kernel_info.hpp:33:
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/include/miopen/kernel.hpp:53:
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/include/miopen/hipoc_kernel.hpp:34:
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/include/miopen/op_kernel_args.hpp:8:
In file included from /usr/include/boost/container/small_vector.hpp:27:
/usr/include/boost/container/vector.hpp:2754:4: error: use of undeclared identifier 'noinline'; did you mean 'inline'?
   BOOST_CONTAINER_NOINLINE iterator priv_insert_forward_range_no_capacity
   ^
/usr/include/boost/container/detail/workaround.hpp:116:37: note: expanded from macro 'BOOST_CONTAINER_NOINLINE'
   #define BOOST_CONTAINER_NOINLINE BOOST_NOINLINE
                                    ^
/usr/include/boost/config/detail/suffix.hpp:639:46: note: expanded from macro 'BOOST_NOINLINE'
#      define BOOST_NOINLINE __attribute__ ((__noinline__))
                                             ^
/opt/rocm/hip/../include/hip/amd_detail/host_defines.h:50:37: note: expanded from macro '__noinline__'
#define __noinline__ __attribute__((noinline))
                                    ^
```

and in all of the above cases, the addition of `#include <hip/hip_version.h>` as an additional header in each `.cpp` file with no other change to the code fixes the compilation error.  I did also try adding this header to `hipoc_kernel.hpp` but that did not resolve the issue.

I'm happy to create a PR to address this issue, though I would like to understand why adding the missing `hip_version.h` header to `hipoc_kernel.hpp` does not appear to resolve the issue.

I have checked this against the latest develop commit, and the issue (at least when compiling `find_db.cpp`) is still there.

### PR DESCRIPTION
**Pull request by junliume**
Contineous work based on #1490

Also should fix #1489

### CODE REVIEW

**Comment by junliume** [comment]
@atamazov trying to assign a reviewer role too.

File: src/include/miopen/op_kernel_args.hpp [hunk 1]
```diff
@@ -4,8 +4,11 @@
 #include <type_traits>
 #include <cstdint>
 #include <half.hpp>
-
+#if HIP_PACKAGE_VERSION_FLAT > 5001999999ULL
+#include <miopen/config.h>
```
Comment (junliume) [inline src/include/miopen/op_kernel_args.hpp:8]: For reviewers' information:
This is puzzling workaround but only reproducible on ROCm 5.1.x: even with config.h included build will fail and complain the same thing about noinline.

**Comment by junliume** [comment]
Ping @pfultz2 @JehandadKhan @JP-Ellis @atamazov  for review :)

**Comment by JP-Ellis** [comment]
Sorry, I haven't had much time to check this.

I noticed there's a few error suppressions such as `cppcheck-suppress knownConditionTrueFalse`.  I don't know whether that is standard practice?

Otherwise, it looks all good by me.

**Review by JP-Ellis — APPROVED** [review]

**Comment by junliume** [comment]
> Sorry, I haven't had much time to check this.
> 
> I noticed there's a few error suppressions such as `cppcheck-suppress knownConditionTrueFalse`. I don't know whether that is standard practice?
> 
> Otherwise, it looks all good by me.

Yes, these suppressions are identified on a case-by-case basis. Technically, they should not be in the same PR (one PR for one purpose rule), while they are simple enough to be included along the way. :)

File: src/include/miopen/op_kernel_args.hpp (hunk 1 above)
Comment (atamazov) [reply src/include/miopen/op_kernel_args.hpp:8]: @junliume How can I reproduce the issue? Tried with ROCm 5.1 (miopen-image built from the scratch using our Dockerfile), but there is no problems with original code (even if config.h is not included). Either I am doing something wrongly or this "puzzling W/A" is not necessary.

File: src/include/miopen/handle.hpp [hunk 2]
```diff
@@ -52,7 +52,11 @@
 
 #if MIOPEN_USE_ROCBLAS
 #include <miopen/manage_ptr.hpp>
+#if HIP_PACKAGE_VERSION_FLAT <= 5001999999ULL
```
[the remark this answers is not in the document]
Comment (atamazov) [reply src/include/miopen/handle.hpp:55]: @junliume I think this should be rocBLAS 2.45, according to https://github.com/ROCmSoftwarePlatform/rocBLAS/commit/99528dd09118e1990a27df1ca39ba30341c5f516.

File: include/miopen/config.h.in [hunk 3]
```diff
@@ -118,3 +118,9 @@
      HIP_PACKAGE_VERSION_PATCH)
 
 #endif
+
+/// Workaround for https://github.com/boostorg/config/issues/392#issuecomment-1109889533
+/// TODO: Remove the W/A as soon we switch to the properly fixed boost.
+#if MIOPEN_BACKEND_HIP
+#include <hip/hip_version.h>
+#endif
```
[the remark this answers is not in the document]
Comment (junliume) [reply include/miopen/config.h.in:126]: @atamazov  no and that's an error. I will fix in a follow up PR.

File: src/include/miopen/op_kernel_args.hpp (hunk 1 above)
Comment (junliume) [reply src/include/miopen/op_kernel_args.hpp:8]: @atamazov  it only appears in our CI and I have had a hard time trying to reproduce it too.
Here is the log
[before_57340fff.log](https://github.com/ROCmSoftwarePlatform/MIOpen/files/8834720/before_57340fff.log)

Comment (atamazov) [reply src/include/miopen/op_kernel_args.hpp:8]: @junliume Thanks, now I see how to reproduce it (-k analyze)

Comment (atamazov) [reply src/include/miopen/op_kernel_args.hpp:8]: @junliume The reason is as follows. The fix in boost works at compile time, when include/boost/config/detail/suffix.hpp is first included. HIP_VERSION must be defined at the moment of first inclusion of that header. Error message says:
```
[ 47%] Building CXX object speedtests/CMakeFiles/speedtest_sequences.dir/sequences.cpp.o
In file included from /var/jenkins/workspace/MLLibs_MIOpen_boost_update/speedtests/sequences.cpp:4:
In file included from /var/jenkins/workspace/MLLibs_MIOpen_boost_update/speedtests/../test/driver.hpp:28:
...
In file included from /var/jenkins/workspace/MLLibs_MIOpen_boost_update/src/include/miopen/op_kernel_args.hpp:8:
In file included from /usr/local/include/boost/container/small_vector.hpp:27:
/usr/local/include/boost/container/vector.hpp:2753:4: error: use of undeclared identifier 'noinline'; did you mean 'inline'?
   BOOST_CONTAINER_NOINLINE iterator priv_insert_forward_range_no_capacity
   ^
```
We can see that `sequences.cpp` includes `driver.hpp`, then `driver.hpp` includes `get_handle.hpp` and so on, and then build failure happens somewhere in depths (`op_kernel_args.hpp`). However `sequences.cpp` includes some other files **_before_** `driver.hpp`, specifically, `miopen/sequences.hpp` and `miopen/rank.hpp`. 

`miopen/sequences.hpp` includes some boost beaders (`boost/range/algorithm/find.hpp`). At that moment config.h is not included yet, and `BOOST_NOINLINE` is defined incorrectly. I will provide a fix.

Comment (atamazov) [reply src/include/miopen/op_kernel_args.hpp:8]: @junliume Done in #1581

File: src/include/miopen/handle.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (junliume) [reply src/include/miopen/handle.hpp:55]: @atamazov  please feel free to make changes in https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1580

**Pull request merged**